patches/api/0387-Paper-Plugins.patch

```diff
@@ -0,0 +1,963 @@
++
++        this.logger = pluginProvider.getLogger();
++    }
++
++    @Override
++    public URL getResource(String name) {
++        return this.findResource(name);
++    }
++
++    @Override
++    public Enumeration<URL> getResources(String name) throws IOException {
++        return this.findResources(name);
++    }
++
++    @Override
++    protected Class<?> loadClass(String name, boolean resolve) throws ClassNotFoundException {
++        return this.loadClass(name, resolve, true);
++    }
++
++    // We do not use the library loader
++    public Class<?> loadClass(@NotNull String name, boolean resolve, boolean checkGlobal) throws ClassNotFoundException {
++        try {
++            Class<?> result = super.loadClass(name, resolve);
++
++            // SPIGOT-6749: Library classes will appear in the above, but we don't want to return them to other plugins
++            if (checkGlobal || result.getClassLoader() == this) {
++                return result;
++            }
++        } catch (ClassNotFoundException ex) {
++        }
++
++        if (checkGlobal) {
++            // This ignores the libraries of other plugins, unless they are transitive dependencies.
++            Class<?> result = this.loader.getClassByName(name, resolve, this.pluginConfiguration, this);  // Paper - prioritize self
++
++            if (result != null) {
++                // If the class was loaded from a library instead of a PluginClassLoader, we can assume that its associated plugin is a transitive dependency and can therefore skip this check.
++                if (result.getClassLoader() instanceof PaperPluginClassLoader existingClassLoader) {
++                    PluginConfiguration classLoaderConfig = existingClassLoader.pluginConfiguration;
++
++                    String identifier = classLoaderConfig.getIdentifier();
++                    if (classLoaderConfig != this.pluginConfiguration && !this.seenIllegalAccess.contains(identifier) && !this.pluginConfiguration.isTransitiveDependency(classLoaderConfig)) {
++                        this.seenIllegalAccess.add(identifier);
++                        if (this.pluginProvider != null) {
++                            this.logger.warn("Loaded class {} from {} which is not a depend or softdepend of this plugin.", name, classLoaderConfig.getDisplayName());
++                        }
++                    }
++                }
++
++                return result;
++            }
++        }
++
++        throw new ClassNotFoundException(name);
++    }
++
++    @Override
++    protected Class<?> findClass(String name) throws ClassNotFoundException {
++        if (name.startsWith("org.bukkit.") || name.startsWith("net.minecraft.")) {
++            throw new ClassNotFoundException(name);
++        }
++        Class<?> result = this.loadedClasses.get(name);
++
++        if (result == null) {
++            String path = name.replace('.', '/').concat(".class");
```

Comment:
Don't use concat just use a simple plus operator there are no concern about minor optimization after a full replace of char

Reply:
Code is based off an internal implementation, there will be some comments explaining this later.